File: backend/app/src/main/kotlin/net/animeta/backend/service/RecommendationService.kt
package net.animeta.backend.service

import com.google.common.cache.Cache
import com.google.common.cache.CacheBuilder
import net.animeta.backend.dto.Credit
import net.animeta.backend.dto.CreditType
import net.animeta.backend.dto.Recommendation
import net.animeta.backend.dto.WorkCredit
import net.animeta.backend.model.User
import net.animeta.backend.repository.WorkStaffRepository
import org.springframework.stereotype.Service
import java.util.concurrent.TimeUnit

@Service
class RecommendationService(
    private val workStaffRepository: WorkStaffRepository
) {
    private val scoreByCreditType = mapOf(
        CreditType.ORIGINAL_WORK to 10,
        CreditType.CHIEF_DIRECTOR to 20,
        CreditType.SERIES_DIRECTOR to 20,
        CreditType.DIRECTOR to 20,
        CreditType.SERIES_COMPOSITION to 10,
        CreditType.CHARACTER_DESIGN to 6,
        CreditType.MUSIC to 5
    )
    private val compatibleCreditTypes = listOf(
        listOf(CreditType.CHIEF_DIRECTOR, CreditType.SERIES_DIRECTOR, CreditType.DIRECTOR),
        listOf(CreditType.SERIES_COMPOSITION, CreditType.ORIGINAL_WORK)
    )
    private val creditCache: Cache<Int, List<Credit>> = CacheBuilder.newBuilder()
        .expireAfterWrite(1, TimeUnit.HOURS)
        .build()

    data class Context(val relatedStaffs: Map<Int, List<WorkStaffRepository.RelatedRow>>)

    fun invalidateCache() {
        creditCache.invalidateAll()
    }

    fun createContext(workIds: List<Int>, user: User): Context {
        if (workIds.isEmpty()) {
            return Context(relatedStaffs = emptyMap())
        }
        val relatedStaffs = workStaffRepository.batchFindRelatedByUser(workIds, user)
            .groupBy { it.personId }
        return Context(relatedStaffs = relatedStaffs)
    }

    fun generate(workId: Int, context: Context): Pair<List<Recommendation>, Int> {
        val workCredits = creditCache.get(workId) {
            workStaffRepository.findByWorkId(workId).mapNotNull { staff ->
                CreditType.fromTask(staff.task)?.let { type ->
                    Credit(
                        type = type,
                        name = staff.person.name,
                        personId = staff.person.id!!
                    )
                }
            }.sortedBy { it.type.ordinal }
        }
        val result = workCredits.mapNotNull { credit ->
            context.relatedStaffs[credit.personId]?.let { staffs ->
                val related = staffs.withIndex().mapNotNull { (index, it) ->
                    // TODO: group same credit type (in FE?)
                    val creditType = CreditType.fromTask(it.staffTask.toLowerCase())
                    if (creditType != null && isCompatible(creditType, credit.type)) {
                        Pair(Pair(creditType.ordinal, index),
                            WorkCredit(it.workId, it.workTitle, creditType))
                    } else {
                        null
                    }
                }.sortedWith(compareBy({ it.first.first }, { it.first.second })).map { it.second }
                Recommendation.ByCredit(
                    credit = credit,
                    related = related.take(2),
                    score = (scoreByCreditType[credit.type] ?: 1) * related.size
                )
            }
        }.filter { it.related.isNotEmpty() }
        val score = result.sumBy {
            if (it is Recommendation.ByCredit) it.score else 0
        } * result.size
        return Pair(result, score)
    }

    private fun isCompatible(a: CreditType, b: CreditType): Boolean {
        if (a == b) {
            return true
        }
        return compatibleCreditTypes.any { it.contains(a) && it.contains(b) }
    }
}

File: backend-ts/src/services/recommendation.service.ts
import { CACHE_MANAGER, Inject, Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Cache } from "cache-manager";
import { sortBy } from "lodash";
import { Credit, CreditType, Recommendation } from "shared/types_generated";
import { Record } from "src/entities/record.entity";
import { StatusType } from "src/entities/status_type";
import { User } from "src/entities/user.entity";
import { WorkStaff } from "src/entities/work_staff.entity";
import { Repository } from "typeorm";

type RelatedStaffRow = {
  personId: number;
  staffTask: string;
  workId: number;
  workTitle: string;
}

export type RecommendationContext = {
  relatedStaffs: Map<number, RelatedStaffRow[]>;
}

@Injectable()
export class RecommendationService {
  private readonly scoreByCreditType: {[P in CreditType]: number} = {
    ORIGINAL_WORK: 10,
    CHIEF_DIRECTOR: 20,
    SERIES_DIRECTOR: 20,
    DIRECTOR: 20,
    SERIES_COMPOSITION: 10,
    CHARACTER_DESIGN: 6,
    MUSIC: 5,
  }
  private readonly compatibleCreditTypes: ReadonlyArray<ReadonlyArray<CreditType>> = [
    ['CHIEF_DIRECTOR', 'SERIES_DIRECTOR', 'DIRECTOR'],
    ['SERIES_COMPOSITION', 'ORIGINAL_WORK'],
  ]
  private readonly taskToCreditType: {[key: string]: CreditType} = {
    "chief director": 'CHIEF_DIRECTOR',
    "series director": 'SERIES_DIRECTOR',
    "director": 'DIRECTOR',
    "character design": 'CHARACTER_DESIGN',
    "animation character design": 'CHARACTER_DESIGN',
    "music": 'MUSIC',
    "series composition": 'SERIES_COMPOSITION',
    "original creator": 'ORIGINAL_WORK',
    "original work": 'ORIGINAL_WORK',
    "original story": 'ORIGINAL_WORK',
    "original manga": 'ORIGINAL_WORK',
  }
  private readonly creditTypeOrder: CreditType[] = [
    'ORIGINAL_WORK',
    'CHIEF_DIRECTOR',
    'SERIES_DIRECTOR',
    'DIRECTOR',
    'SERIES_COMPOSITION',
    'CHARACTER_DESIGN',
    'MUSIC',
  ]
  
  constructor(
    @InjectRepository(WorkStaff) private workStaffRepository: Repository<WorkStaff>,
    @Inject(CACHE_MANAGER) private cache: Cache,
  ) {}

  async createContext(workIds: number[], user: User): Promise<RecommendationContext> {
    const relatedStaffs = new Map<number, RelatedStaffRow[]>()
    if (workIds.length > 0) {
      const relatedStaffRows = await this.workStaffRepository.createQueryBuilder('s')
        .select('s2.person_id', 'personId')
        .addSelect('s2.task', 'staffTask')
        .addSelect('s2.work_id', 'workId')
        .addSelect('r.title', 'workTitle')
        .innerJoin(WorkStaff, 's2', 's2.person_id = s.person_id AND s2.work_id <> s.work_id')
        .innerJoin(Record, 'r', 'r.work_id = s2.work_id')
        .where('s.work_id IN (:...workIds) AND s2.work_id NOT IN (:...workIds)', { workIds })
        .andWhere('r.user_id = :userId', {userId: user.id})
        .andWhere('r.status_type IN (:...statusTypes)', {statusTypes: [StatusType.WATCHING, StatusType.FINISHED]})
        .groupBy('s2.person_id')
        .addGroupBy('s2.task')
        .addGroupBy('s2.work_id')
        .addGroupBy('r.title')
        .addGroupBy('r.updated_at')
        .orderBy('r.updated_at', 'DESC')
        .getRawMany<RelatedStaffRow>()
      for (const row of relatedStaffRows) {
        const group = relatedStaffs.get(row.personId) ?? []
        relatedStaffs.set(row.personId, group.concat(row))
      }
    }
    return {relatedStaffs}
  }

  async generate(workId: number, context: RecommendationContext): Promise<{
    recommendations: Recommendation[],
    recommendationScore: number
  }> {
    const workCredits = await this.cache.wrap<Credit[]>(`credit:${workId}`, async () => {
      return (await this.workStaffRepository.find({ relations: ['person'], where: {work_id: workId} }))
        .map(staff => {
          const creditType = this.taskToCreditType[staff.task.toLowerCase()]
          if (!creditType) return null
          return {
            type: creditType,
            name: staff.person.name,
            personId: staff.person.id,
          }
        })
        .filter(it => it).map(it => it!)
        .sort((a, b) => this.creditTypeOrder.indexOf(a.type) - this.creditTypeOrder.indexOf(b.type))
    }, { ttl: 60 * 60 })
    const result = workCredits.map(credit => {
      const staffs = context.relatedStaffs.get(credit.personId)
      if (!staffs) return null
      let related0 = staffs.map((it, index) => {
        const creditType = this.taskToCreditType[it.staffTask.toLowerCase()]
        return creditType && this.isCompatible(creditType, credit.type) ?
          [[this.creditTypeOrder.indexOf(creditType), index],
            { workId: it.workId, workTitle: it.workTitle, creditType }]
          : null
      })
        .filter(it => it).map(it => it!)
      related0 = sortBy(related0, ([score, ]) => score)
      const related = related0.map(([, value]) => value)
      return {
        credit,
        related: related.slice(0, 2),
        score: (this.scoreByCreditType[credit.type] ?? 1) * related.length,
      }
    }).filter(it => it && it.related.length > 0).map(it => it!)
    const score = result.reduce((acc, it) => acc + it.score, 0) * result.length
    return {
      recommendations: result,
      recommendationScore: score,
    }
  }

  private isCompatible(a: CreditType, b: CreditType): boolean {
    if (a === b)
      return true
    return this.compatibleCreditTypes.some(it => it.includes(a) && it.includes(b))
  }
}
